### ISSUE TITLE
Brainstorm ideas for making the diagram more clear & intuitive to understand

### ISSUE THREAD

**Issue by keyserj** [amelioro/ameliorate#554]
### Describe your issue

The diagram is currently not trivial to follow, presumably because some aspects aren't intuitive and because there's a lot of clutter, making it hard to know where to look.

Awkward things:
- Add Node Buttons create a lot of clutter when a node is selected/hovered
- Panes #551
- Problem/Solution nodes don't really stand out #552
- Edges are ambiguous/confusing if you aren't used to reading them
  - reading bottom to top is a little unexpected
  - reading direction isn't clear enough from the small arrows (maybe label box could be shaped like an arrow?)

### Additional context

Related #541 #478 #548 #434 #298

**Comment by drared** [comment]
I think this issue is still relevant. I spent some time looking at the 'cars too fast' demo and concluded it was not as clear as one would want. 
I felt if I was to start from scratch and try to make a diagram to describe that situation, I would go with a different layout. Some of the nodes end up in positions that do not reflect their impact. I had trouble following the edges when there were multiple edges, close together and flowing behind other nodes. 
Not being able to move the nodes or group the nodes felt somewhat restrictive. I was wondering if the different views would address this but it appeared I could access that demo model to explore those alternative views.

**Comment by keyserj** [comment]
@drared thanks for the feedback!

> I would go with a different layout

I'd enjoy discussing better layout ideas - did you have specific placement ideas for the [cars-too-fast](https://ameliorate.app/examples/detailed-cars-going-too-fast) example? There are some limitations with the layout library I'm using, but it does have a lot of features too.

> trouble following the edges when there were multiple edges, close together and flowing behind other nodes

Yeah I agree, a lot of edges are hard to follow. The layout algorithm tries to minimize the number of edge crossings but it's not perfect. It also doesn't avoid edge overlap with nodes at all. There is an "avoid edge overlap" option I added, but it kind of makes the overall layout worse ([details here](https://github.com/amelioro/ameliorate/issues/559#issuecomment-2496393356) - basically it'd be ideal if the layout library makes a change).

> Could adding identifying features to the edges (ie color, thickness, shading etc) be beneficial?

That's a good idea that I've only briefly experimented with. I also considered using animations to indicate which direction the edges flow in. I'm not sure the best way to indicate some information via decoration without overloading the visuals, and that has a lot of room to explore more.

> Not being able to move the nodes or group the nodes felt somewhat restrictive

I do have a ticket for allowing moving the nodes manually https://github.com/amelioro/ameliorate/issues/196 that hopefully would help a little. Grouping nodes seems like a good way to help the automatic layout, but I haven't explored that.

> I was wondering if the different views would address this but it appeared I could access that demo model to explore those alternative views.

Hmm I'm not sure what you mean by this? You should be able to switch views via the Quick Views dropdown (at the top of screen for desktop):

![Image](https://github.com/user-attachments/assets/3915bac9-1a31-48e9-9129-3b9ab506bbcf)

Somewhat related: I also just added a few prototype ideas for viewing the topic information _without_ a diagram here https://github.com/amelioro/ameliorate/discussions/541#discussioncomment-12706593 and below that comment. I think maybe a diagram is sometimes harder to navigate on mobile or for new users. Though I do still hope to make many improvements to the diagram to make it friendlier.

**Comment by drared** [comment]
@keyserj, in regards to layout... this must be a tough challenge for ameliorate as different users and different topics would probably benefit from different layouts. Hopefully it will be condensed to something manageable. I will draw a rough diagram on how I would naturally draw a diagram to represent this situation. Before I do that though, do you have an online mind mapping tool you use? Mindmup, miro or the like? It might pay off to put ideas on a platform that can be edited?

It occurs to me when reading your response in regards to defining edges, a hover effect would probably be all that is needed. We tend to move the mouse over where we are focused and a hover effect would show the connections we are interested in. 

Thanks for pointing out the example. I was going off the youtube video and when I clicked the first link in the description, it went to 'play' and so I wasn't sure I could access that example. You just had to say 'huh' for me to know I needed to look a bit deeper :)

I will check the prototype later today (it's almost midday here). I am betting on diagrams as well. I think if you keep your eye on the prize (whatever that is) and keep working at it, it will happen. If there was a leaderboard for argument maps (even though I know that is not your aim), ameliorate is right up there. 

I applaud your willingness to receive feedback. To me, it's a key factor for a successful product.

**Comment by drared** [comment]
I spent time looking at how I would stick this in a diagram and I concluded I would of gone with a table.  For me, I would compress benefits, detriments & obstacles into criteria for your solutions. A complicating factor is that each solution will affect each cause differently. Hence I would want it to become another dimension which complicates a 2d chart. I could imagine ways to express this but my skills with the diagram app makes it a time consuming project. 

I ended up finding miro to be too restrictive to use. Too costly a subscription to just have something that I might use from time to time. I ended up playing with draw.io which seems like a great fit. 

I will note some thoughts I had during the exercise...
I wanted some quantification of the importance of the problem. How much of a concern are the speeding cars? Has a budget been allocated? This type of information can impact on the solution assessment. Along the same lines, if a 'hygiene' factor existed (something that has to be there) that ruled  out as a particular solution, how would that information be captured? For example, the budget is 20,000 while the cost of traffic lights was 50,000, the diagram should indicate that was not an appropriate solution. 

I find the positioning of the 'causes' is missing an opportunity to convey some meaning. If some causes have a disproportionate impact on the problem, that could be indicated by the position. I would go with top to bottom orientation but left to right could work. Visual cues could be used like larger squares, different colours and the like. As is, each cause is equal and that is not likely. 

I see multiple components which I would describe as criteria to find the most appropriate solution. They are separated out when I would think they should be applied equally. Equally so, in this case, for detriments and obstacles. They all could easily be described and applied to all solutions equally. To clarify my point, I see detriment, 'significantly reduces traffic throughput' applied to stop sign when I would think that same criteria should be applied to all solutions. I understand that is the point of a dynamic model like this, I can adjust that, but a table is what we generally use for such comparisons. 

To describe a diagram that I think would enlighten users: It would show a table type comparison for each solution against each criteria (a combination of detriement, component, benefit & obstacle), and this would be done for each cause. They could feed into a single table showing the most suitable solution based on the scores and the weighting given by each cause.  I can see it clearly but I am sure this description would be a pain to follow so I have done a rough sketch of what I am imagining. 

So yeah, unfortunately, for this example, after some thought, I found a table format was more appropriate. I need to have a think about when the other components would be used. 

[Ameliorate example_20250404_130338.pdf](https://github.com/user-attachments/files/19597089/Ameliorate.example_20250404_130338.pdf)

**Comment by keyserj** [comment]
@drared thanks for thinking about this thoroughly! The image is helpful.

Interesting that you ended up going with a table. Ameliorate actually does support criteria tables, but only for comparing how solutions address a problem, with no special logic for combining how they solve multiple sub problems. This specific topic has a table here https://ameliorate.app/examples/detailed-cars-going-too-fast?view=Tradeoff+table.

I hadn't considered using the table for evaluating how well solutions address specific causes! That seems nice, perhaps even more so when a root problem gets more complex.

In my experience, the table is best when you already know what options you have, and you already understand the details/impacts of the options - while the causal nature of the diagram makes it preferable to the table when wanting to understand the problem better (causes/effects), brainstorm potential solutions, or think through/flesh out their details/impacts. I've been thinking that the diagram is ideal to start with, then the table is best when you understand enough to make a decision (and basically each benefit/detriment would map to a criterion). Do you think you'd feel that the table is still best when the problem or solution causes/effects are less clear?

> How much of a concern are the speeding cars?

I might communicate this by scoring the problem's detriments, or by adding statistics about accidents or something via fact nodes.

> Has a budget been allocated?

Hmm this is an important thing to be able to capture. Right now I guess I'd probably use a comment or fact node, but I think it'd be best to have first class support for something like "implementation status" or "previous implementation attempt details".

> For example, the budget is 20,000 while the cost of traffic lights was 50,000, the diagram should indicate that was not an appropriate solution.

Yeah Ameliorate still needs something like being able to make hard requirements.

> As is, each cause is equal and that is not likely.

If you show scores, the score colors can help distinguish the importance of these. I've considered sorting left to right (when diagram is flowing vertically) by score, but it can make the layout look less balanced - I think one of the tile based/mobile views I've been prototyping would be much better for easily sorting. Other ways to communicate importance could be good too.

As an aside, I think Miro should be free for only time-to-time usage? But in any case I think draw.io is good too!

Another aside - some of my comments are probably better illustrated via images. Can add images later if that would help.

**Comment by drared** [comment]
Considering the various arrangements of the structure was a good exercise for me to go through. I approached it with the idea that it should be better than a paragraph describing the situation. I am thinking the aim of the diagram is for someone to be able to look at the diagram and relatively quickly see the key factors weighing on the situation. For me, that wasn't the case with how it is. I needed to move around the diagram and put it all together. Some of the key aspects, which I would expect to be grouped, were on their own. Of course, it gets better the more you look at it but that is not the aim right? 

I had seen the tradeoff table. It was only when I started considering how I would arrange the nodes that I realised they resembled a table. 
I agree with you that the table look I ended up with would constrain brainstorming and information gathering. But then, at some point, it needs to be able to sorted/ordered for people to make better sense of the issue. I am not sure if what I noticed in this example, with various benefits, detriements and obstacles being able to be grouped would apply universally. We would need to look at a wide variety of topics to determine that.

OK, so I am thinking I might of approached Ameliorate with a different view to what I should of. I didn't approach it as a tool to quickly add the different aspects for a situation with many factors. I guess I would need to consider a different topic for that. Still, if the idea is to finally have a diagram that summed up the topic efficiently, Some new tools might be needed to allow summarizing, gathering, restructuring or the like? That would all be another level of difficulty I imagine?

I will have a go at adding a new topic when I'm able to.

**Comment by keyserj** [comment]
> I am thinking the aim of the diagram is for someone to be able to look at the diagram and relatively quickly see the key factors weighing on the situation. For me, that wasn't the case with how it is.

> Some new tools might be needed to allow summarizing, gathering, restructuring or the like?

Yeah being able to see the key factors quickly, to drill down where you want from there, to restructure easily, these are really important for Ameliorate. And I agree that more features/improvements need to be made to make Ameliorate better at these, but with continual efforts I think the tool can go far!

> I didn't approach it as a tool to quickly add the different aspects for a situation

Yeah there's the viewing side of it, and the editing/building side of it.

> I will have a go at adding a new topic when I'm able to.

Sounds great, let me know if you have any questions.
